### PR TITLE
[ONNX] Fix bugs for the `onnx.OneHot` operator

### PR DESCRIPTION
**Pull request by angelz913**
This commit fixes the bugs for the `onnx.OneHot` operator by:

1) Converting negative indices to non-negative indices
2) Handling both `int` and `float` types for `off` and `on` values
3) Using the correct result type

It also includes a new unit test.

### CODE REVIEW

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 1]
```diff
@@ -1654,13 +1654,22 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
           depth = rewriter.create<Torch::AtenIntScalarOp>(
               loc, rewriter.getType<Torch::IntType>(), depth);
 
-        auto selectTy = rewriter.getType<Torch::ValueTensorType>(
-            llvm::SmallVector<int64_t>{1}, valuesTy.getDtype());
-
+        // Apply bounds checking on the indices
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1657]: Would be easier to understand if this comment is either more concise or more specific. E.g. I personally would have written something like `// If indices < 0, indices += depth`

another perhaps better option is to remove this comment entirely - - the code is decently self-explanatory

**Review by renxida — APPROVED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -1654,13 +1654,22 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+        Type boolTy = rewriter.getType<Torch::ValueTensorType>(
+            indicesTy.getSizes(), rewriter.getI1Type());
         Value zero = rewriter.create<Torch::ConstantIntOp>(
-            loc, rewriter.getI64IntegerAttr(0));
+            loc, intTy, rewriter.getI64IntegerAttr(0));
         Value one = rewriter.create<Torch::ConstantIntOp>(
-            loc, rewriter.getI64IntegerAttr(1));
+            loc, intTy, rewriter.getI64IntegerAttr(1));
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1663]: You don't need to specify the `intTy` here.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 3]
```diff
@@ -1712,7 +1721,7 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
 
         onehotTy = rewriter.getType<Torch::ValueTensorType>(
             onehotShape, resultType.getDtype());
-        onehot = rewriter.create<Torch::AtenWhereScalarOp>(loc, onehotTy,
+        onehot = rewriter.create<Torch::AtenWhereScalarOp>(loc, resultType,
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1724]: I'm not sure why the `onehotTy` was created, so can't comment on removing it either. It's better if you confirm this with @rsuderman before merging.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 3 above)
Comment (angelz913) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1724]: This is addressed in PR #3333

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 1 above)
Comment (angelz913) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: Removed this comment

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (angelz913) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: Removed `intTy`

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Pull request merged**